modules/aws-http4s/src/smithy4s/aws/internals/AwsSigning.scala

```diff
@@ -106,79 +116,78 @@ private[aws] object AwsSigning {
-          `Host` -> request.uri.host.map(_.renderString).orNull,
-          `X-Amz-Date` -> timestamp.conciseDateTime,
-          `X-Amz-Security-Token` -> credentials.sessionToken.orNull,
-          `X-Amz-Target` -> (serviceName + "." + operationName)
-        ).filterNot(_._2 == null)
-
-        val canonicalHeadersString = baseHeadersList
-          .map { case (key, value) =>
-            key.toString.toLowerCase + ":" + value.trim
-          }
-          .mkString(newline)
-        lazy val signedHeadersString = baseHeadersList.map(_._1).map(_.toString.toLowerCase()).mkString(";")
-
-        val payloadHash = sha256HexDigest(body.toArray)
-        val pathString = request.uri.path.toAbsolute.renderString
-        val canonicalRequest = new StringBuilder()
-          .append(request.method.name.toUpperCase())
-          .append(newline)
-          .append(pathString)
-          .append(newline)
-          .append(canonicalQueryString)
-          .append(newline)
-          .append(canonicalHeadersString)
-          .append(newline)
-          .append(newline)
-          .append(signedHeadersString)
-          .append(newline)
-          .append(payloadHash)
-          .result()
-
-        val canonicalRequestHash = sha256HexDigest(canonicalRequest)
-        val signatureKey = getSignatureKey(
-          credentials.secretAccessKey,
-          timestamp.conciseDate,
-          region.value,
-          endpointPrefix
-        )
-        val stringToSign = List[String](
-          algorithm,
-          timestamp.conciseDateTime,
-          credentialsScope,
-          canonicalRequestHash
-        ).mkString(newline)
-        val signature = toHexString(hmacSha256(stringToSign, signatureKey))
-        val authHeaderValue = s"${algorithm} Credential=${credentials.accessKeyId}/$credentialsScope, SignedHeaders=$signedHeadersString, Signature=$signature"
-        val authHeader = Headers("Authorization" -> authHeaderValue)
-        val baseHeaders = Headers(baseHeadersList.map { case (k, v) => Header.Raw(k, v) })
-        authHeader ++ baseHeaders
-      }
+    (request: Request[F]) =>
+      preSign(request).flatMap { case (payloadHash, preparedRequest) =>
+        val awsHeadersF = (timestamp, credentials, region).mapN { case (timestamp, credentials, region) =>
+          val credentialsScope = s"${timestamp.conciseDate}/$region/$endpointPrefix/aws4_request"
+          val queryParams: Vector[(String, String)] =
+            request.uri.query.toVector.sorted.map { case (k, v) => k -> v.getOrElse("") }
+          val canonicalQueryString =
+            if (queryParams.isEmpty) ""
+            else
+              queryParams
+                .map { case (k, v) =>
+                  URLEncoder.encode(k, StandardCharsets.UTF_8.name()) + "=" + URLEncoder.encode(v, StandardCharsets.UTF_8.name())
+                }
+                .mkString("&")
+
+          // // !\ Important: these must remain in the same order
+          val baseHeadersList = List(
```

Comment:
**nitpick:** I know this is already the case, but the nulls are making me uncomfortable. Can we move to options?

```suggestion
          val baseHeadersList = List(
            `Content-Type` -> preparedRequest.contentType.map(contentType.value(_))
            ).collect { case (k, Some(v)) => k -> v }
```

instead of the `filterNot(_._2 == null)`

Reply:
fine